simple/msg_data_integrity.c

```diff
@@ -0,0 +1,717 @@
+{
+	int j = 0;
+	printf("%s = ", buffer);
+	for(j=0; j<size; j++) {
+		printf("%c", array[j]);
+        }
+        printf("\n");
+}
+
+static int send_xfer(int size, int iter)
+{
+	struct fi_cq_entry comp;
+	int ret;
+	int j;
+	int msg_index;
+	char *msg_buf;
+
+	if (iter >= 0) {
+		msg_index = (iter*common_seed) % msg_char_length;
+	} else {
+		msg_index = 0;
+	}
+	while (!credits) {
+		ret = fi_cq_read(scq, &comp, 1);
+		if (ret > 0) {
+			goto post;
+		} else if (ret < 0 && ret != -FI_EAGAIN) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(scq, "scq");
+			} else {
+				FT_PRINTERR("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	}
+
+	credits--;
+post:
+	msg_buf = (char *)send_buf;
+
+	for (j = 0; j < size; j++) {
+		msg_buf[j] = msg_char[msg_index++];
+		if (msg_index >= msg_char_length) {
+			msg_index = 0;
+		}
+	}
+
+	ret = fi_send(ep, send_buf, (size_t) size, fi_mr_desc(send_mr), 0, NULL);
+	if (ret)
+		FT_PRINTERR("fi_send", ret);
+
+
+	if (detailed_output) {
+		print_byte_array("Send Buffer", msg_buf, size);
+	}
+
+	return ret;
+}
+
+static int recv_xfer(int size, int iter)
+{
+	struct fi_cq_entry comp;
+	char *recv_data;
+	char *expected_data;
+	int msg_index;
+	int ret;
+	int j = 0;
+
+	if (iter >= 0) {
+                msg_index = (iter*common_seed) % msg_char_length;
+        } else {
+                msg_index = 0;
+        }
+
+	expected_data = (char *) malloc(size);
+        if (!expected_data) {
+                perror("malloc");
+                return -1;
+        }
+
+        for (j = 0; j < size; j++) {
+                expected_data[j] = msg_char[msg_index++];
+                if (msg_index >= msg_char_length) {
+                        msg_index = 0;
+                }
+        }
+
+	do {
+		ret = fi_cq_read(rcq, &comp, 1);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			if (ret == -FI_EAVAIL) {
+				cq_readerr(rcq, "rcq");
+			} else {
+				FT_PRINTERR("fi_cq_read", ret);
+			}
+			return ret;
+		}
+	} while (ret == -FI_EAGAIN);
+
+	recv_data = (char *)recv_buf;
+
+	for(j = 0; j < size; j++) {
+		if (expected_data[j] == recv_data[j]) {
+			continue;
+		}
+		first_error_byte = j;
+		break;
+	}
+	if (first_error_byte == -1) {
+		goto no_error;
+	}
+	if (iter < 0) {
+		printf("Test Name = %s, Sync test. Size = %d First_Error_byte = %d\n", 
+			test_name, size, first_error_byte);
+	} else {
+		printf("Test Name = %s, Iteration = %d Size = %d"
+			" First_Error_byte = %d\n", test_name, iter,
+			size, first_error_byte);
+	}
+	first_error_byte = -1;
+
+	if (allowed_error_count > 0) {
+		error_count++;
+	}
+	if (detailed_error) {
+                print_error = 1;
+        }
+
+no_error:
+	if (detailed_output || print_error) {
+		print_byte_array("Expd Buffer", expected_data, size);
+		print_byte_array("Recv Buffer", recv_data, size);
+	}
+
+	memset(recv_buf, 0, buffer_size);
+	if (iter < 0) {
+		next_size = transfer_size;
+	}
+	ret = fi_recv(ep, recv_buf, next_size, fi_mr_desc(recv_mr), 0, recv_buf);
```

Comment:
do you still have an issue where the last recv posted here matches up with the recv posted in `ft_finalize`?